quickwit/quickwit-indexing/src/lib.rs

```diff
@@ -98,3 +104,43 @@ pub async fn start_indexing_service(
     }
     Ok(indexing_service)
 }
+
+trait InstrumentMetric: Sized {
+    fn instrument_waiting_time(self, labels: &[&str]) -> InstrumentedMetric<Self> {
+        let histogram = INDEXER_METRICS
+            .message_waiting_time
+            .with_label_values(labels);
+        let histogram_timer = histogram.start_timer();
+        InstrumentedMetric {
+            inner: self,
+            histogram_timer: Some(histogram_timer),
+        }
+    }
+}
+
+pin_project! {
+    pub struct InstrumentedMetric<T> {
+        #[pin]
+        inner: T,
+        #[pin]
+        histogram_timer: Option<HistogramTimer>,
+    }
+}
+
+impl<T: Future> Future for InstrumentedMetric<T> {
+    type Output = T::Output;
+
+    fn poll(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Self::Output> {
+        let mut this = self.project();
+        let output = match this.inner.poll(cx) {
+            Poll::Pending => return Poll::Pending,
+            Poll::Ready(output) => output,
+        };
+        if let Some(histogram_timer) = this.histogram_timer.take() {
+            histogram_timer.observe_duration();
+        };
+        Poll::Ready(output)
+    }
+}
+
+impl<T: Sized> InstrumentMetric for T {}
```

Comment:
If it compiles, it might be worth making it 

```suggestion
impl<T: Future + Sized> InstrumentMetric for T {}
```

Hopefully it might prevent silly autocomplete on all objects.